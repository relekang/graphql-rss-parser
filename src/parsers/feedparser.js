const Readable = require('stream').Readable
const FeedParser = require('feedparser')

const { ParserError, NotAFeedError } = require('../errors')

module.exports = function parseString (feed) {
  return new Promise((resolve, reject) => {
    try {
      const feedparser = new FeedParser()
      feedparser.on('error', error => {
        reject(new ParserError(error))
      })

      let parsedFeed
      feedparser.on('readable', function () {
        const meta = this.meta
        if (!parsedFeed) {
          parsedFeed = Object.assign({}, meta, { entries: [], parser: 'FEEDPARSER' })
        }

        let item
        while ((item = this.read())) {
          delete item.meta
          parsedFeed.entries.push(Object.assign({}, item, {pubDate: new Date(item.pubDate).toISOString()}))
        }
      })

      feedparser.on('end', function () {
        resolve(parsedFeed);
      })

      const stream = new Readable()
      stream.pipe(feedparser)
      stream.push(feed)
      stream.push(null)
    } catch (error) {
      console.log({error})
      reject(error)
    }
  })
  .catch(error => {
    if (error.message === 'Not a feed') {
      throw new NotAFeedError(error)
    }
    throw error
  })
}
