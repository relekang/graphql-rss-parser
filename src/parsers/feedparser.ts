import { Readable } from 'stream'
import FeedParser from 'feedparser'
import _debug from 'debug'

import { EmptyParserOutputError, NotAFeedError, ParserError } from '../errors'
import { Item, ParserResponse } from '../types'

const debug = _debug('graphql-rss-parser:parsers:feedparser')

export function parse(feed: string): Promise<ParserResponse> {
  return new Promise<ParserResponse>((resolve, reject) => {
    try {
      debug('starting to parse')
      const feedparser = new FeedParser({})
      feedparser.on('error', (error: Error) => {
        debug('parsing failed with error', error)
        reject(new ParserError(error, 'FEEDPARSER'))
      })

      let meta: FeedParser.Meta
      const items: Item[] = []
      feedparser.on('readable', function (this: FeedParser) {
        meta = meta || (this.meta as FeedParser.Meta)

        let item
        while ((item = this.read())) {
          items.push({
            title: item.title,
            description: item.description,
            link: item.link,
            categories: item.categories,
            pubDate: item.pubdate ? new Date(item.pubdate).toISOString() : undefined,
            author: item.author,
            guid: item.guid,
          })
        }
      })

      feedparser.on('end', function () {
        debug('done parsing')
        if (!meta) {
          return reject(new EmptyParserOutputError())
        }
        resolve({
          parser: 'FEEDPARSER',
          title: meta.title,
          description: meta.description,
          link: meta.link,
          feedLink: undefined,
          entries: items,
        })
      })

      const stream = new Readable()
      stream.pipe(feedparser)
      stream.push(feed)
      stream.push(null)
    } catch (error) {
      debug('parsing failed with error', error)
      reject(new ParserError(error, 'FEEDPARSER'))
    }
  }).catch((error) => {
    debug('parsing failed with error', error)
    if (error.message === 'Not a feed' || error.cause.message === 'Not a feed') {
      throw new NotAFeedError(error)
    }
    throw error
  })
}
