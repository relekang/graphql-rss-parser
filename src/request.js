const superagent = require('superagent')

const { EmptyHttpResponseError, NotFoundError } = require('./errors')

module.exports = async function request(url) {
  const response = await superagent(url).buffer()

  if (!response.text) throw new EmptyHttpResponseError()
  if (response.notFound) throw new NotFoundError(url)

  return response.text
}
