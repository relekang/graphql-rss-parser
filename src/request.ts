import axios from 'axios';
import _debug from 'debug';

const debug = _debug('graphql-rss-parser:request');

import {
  ConnectionRefusedError,
  DnsLookupError,
  EmptyHttpResponseError,
  UnknownRequestError,
  UpstreamHttpError,
  UpstreamEncryptionError,
  TimeoutError,
} from './errors';

const TIMEOUT = 30 * 1000;

export default async function request(url: string) {
  try {
    debug(`requesting ${url}`);
    const response = await axios({
      url,
      headers: {
        'User-Agent': 'graphql-rss-parser',
      },
      timeout: TIMEOUT,
      responseType: 'arraybuffer',
      transformResponse: undefined,
    });
    debug(`response from ${url} status-code=${response.status}`);
    if (!/2\d\d/.test(response.status.toString())) {
      throw new UpstreamHttpError('Not found', response.status);
    }
    if (!response.data) {
      throw new EmptyHttpResponseError();
    }
    return {
      text: response.data.toString(),
      status: response.status,
      contentType: response.headers['content-type'],
      headers: response.headers,
    };
  } catch (error: any) {
    debug(`request to ${url} failed with error`, error);
    if (error.response && error.response.status) {
      throw new UpstreamHttpError('Upstream HTTP error', error.response.status);
    }
    if (error.code === 'ENOTFOUND' || error.code === 'EAI_AGAIN') {
      throw new DnsLookupError();
    }
    if (error.code === 'ECONNREFUSED' || error.code === 'ECONNRESET') {
      throw new ConnectionRefusedError();
    }

    if (error.code === 'ECONNABORTED' || error.code === 'ETIMEDOUT') {
      throw new TimeoutError();
    }

    if (
      error.constructor === EmptyHttpResponseError ||
      error.constructor === UpstreamHttpError
    ) {
      throw error;
    }

    if (/certificate|tls|ssl/.test(error.toString())) {
      throw new UpstreamEncryptionError(error);
    }

    throw new UnknownRequestError(error);
  }
}
