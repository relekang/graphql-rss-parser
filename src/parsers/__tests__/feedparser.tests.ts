/* eslint-env jest */
import { parse } from '../feedparser';
import request from '../../request';

test('should parse string from rolflekang.com/feed.xml', async () => {
  const { text } = await request('https://rolflekang.com/feed.xml');

  expect(await parse(text)).toMatchSnapshot();
});

test('should parse string from google.blogspot.com/feeds/posts/default', async () => {
  const { text } = await request(
    'http://google.blogspot.com/feeds/posts/default'
  );

  expect(await parse(text)).toMatchSnapshot();
});
