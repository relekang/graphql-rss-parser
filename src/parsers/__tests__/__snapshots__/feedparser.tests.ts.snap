// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse string from google.blogspot.com/feeds/posts/default 1`] = `
Object {
  "description": null,
  "entries": Array [
    Object {
      "author": "Brett Wiltshire",
      "categories": Array [],
      "description": "<i>Posted by Brett Wiltshire, Blogger CEO</i><br /><br />This morning we’re beyond thrilled to announce that Blogger has signed a definitive agreement to be acquired by Google, the Internet search company. This is exciting news not only for all of us on the Blogger team, but for our users, our partners, and most importantly -- the blogosphere itself. <br /><br />Understandably, you probably have lots of questions about what this means for Blogger and Blogger users. Below, we've put together some initial answers to many of the biggest questions. More info will be available as we figure it out. Thanks for your support as we transfer into this next exciting phase. &nbsp; <br /><br /><br /><b>Q: Why did Blogger sell to Google?</b><br />A: Well, on the surface, it may look obvious: A company of Google's size could give Blogger the resources we needed to do things better, faster, bigger. It's been a long eleven+ years since we started the company, and not all of them were very fun. We had been making serious progress over the last year or so, but bootstrapping (growing without funding) is always a slow process, and we were a long way from where we wanted to be. We wanted to offer a better service. And with Google, it wasn't just their size, of course. They had clearly done an incredible job building their technology and business, and we'd been big fans. <br /><br />However, that doesn't mean it was an easy decision. We'd seen many small companies doing interesting things die (or at least become uninteresting) after getting acquired. It was only after becoming convinced that: a) There were sensible, cool, powerful things we could do on the technology/product side with Google that we couldn't do otherwise; and b) It was a good company, run by people we liked, who wanted the same things we did (i.e., they wouldn't screw up Blogger or make our lives miserable). <br /><br />We became convinced both of these were the case rather quickly after thinking about the product side and talking to lots of Googlers. &nbsp; Also, Google liked our logo. And we liked their food. <br /><br /><b>Q: Will Blogger go away?&nbsp;</b><br />A: Nope. Blogger is going to maintain its branding and services. While we may integrate with Google in certain areas there will always be a Blogger. <br /><br /><b>Q: What does the acquisition mean to Blogger users?&nbsp;</b><br />A: Greater reliability, new innovative products and a whole lot more that we aren't ready to share quite yet (truth is, we're still figuring a lot of it out). Right now, we're mostly bolstering up our hardware -- making things faster and more reliable -- and getting settled. Next, we're going to be working on some killer ideas we've been wanting to implement for years. It'll be fun. <br /><br /><b>Q: Will there be any changes to my account?&nbsp;</b><br />A: Not right now but if anything does change we will notify you ahead of time, as we've done in the past.<br /><br /><b>Q: Will there be any changes to the Blogger products?&nbsp;</b><br />A: We will be making some changes in our product line. We've been working on a new version of Blogger for some time now that will be coming out soon. We'll tell you more as soon as we know. <br /><br /><b>Q: What are your plans for the future?&nbsp;</b><br />A: We are building a nuclear powered... Wait, you almost had us. We aren't telling, yet! But we will have more in a few weeks. <br /><br /><b>Q: Does this mean my blog will rank higher in Google search results?&nbsp;</b><br />A: Nope. It does mean your blog might be stored physically closer to Google but that's about it. The people at Google have done a great job over the years making sure their search results are honest and objective and there's no reason they would change that policy for Blogger or anyone else. <br /><br /><b>Q: What will happen to all the nice kids that work on Blogger?&nbsp;</b><br />A: We’ll still be working on Blogger and making it better. <br /><br /><b>Q: Are you still as handsome as ever?&nbsp;</b><br />A:<br /><div class=\\"separator\\" style=\\"clear: both; text-align: center;\\"><a href=\\"http://1.bp.blogspot.com/_UUTay7LUoq0/TXVACrMZNOI/AAAAAAAAADI/_WWawmSQnIM/s1600/_MG_1538.jpg\\" imageanchor=\\"1\\" style=\\"margin-left: 1em; margin-right: 1em;\\"><img border=\\"0\\" src=\\"http://1.bp.blogspot.com/_UUTay7LUoq0/TXVACrMZNOI/AAAAAAAAADI/_WWawmSQnIM/s400/_MG_1538.jpg\\" width=\\"500\\" /></a></div>",
      "guid": "tag:blogger.com,1999:blog-2120328063286836889.post-7340716506563491347",
      "link": "http://google.blogspot.com/2011/04/google-to-acquire-blogger.html",
      "pubDate": "2011-04-01T15:06:00.000Z",
      "title": "Google to Acquire Blogger",
    },
  ],
  "feedLink": undefined,
  "link": "http://google.blogspot.com/",
  "parser": "FEEDPARSER",
  "title": "Google on BlogSpot",
}
`;

exports[`should parse string from rolflekang.com/feed.xml 1`] = `
Object {
  "description": "undefined",
  "entries": Array [
    Object {
      "author": null,
      "categories": Array [
        "letsencrypt",
        "acme",
        "certbot",
        "ansible",
      ],
      "description": null,
      "guid": "https://rolflekang.com/using-certbot-with-ansible",
      "link": "https://rolflekang.com/using-certbot-with-ansible",
      "pubDate": "2020-11-08T00:00:00.000Z",
      "title": "Using certbot with Ansible",
    },
    Object {
      "author": null,
      "categories": Array [
        "ansible",
        "devops",
        "ops",
        "infrastructure-as-code",
      ],
      "description": null,
      "guid": "https://rolflekang.com/ansible-handlers-in-loops",
      "link": "https://rolflekang.com/ansible-handlers-in-loops",
      "pubDate": "2020-09-12T00:00:00.000Z",
      "title": "Using Ansible handlers in loops",
    },
    Object {
      "author": null,
      "categories": Array [
        "nextjs",
        "javascript",
        "curl",
        "nodejs",
      ],
      "description": null,
      "guid": "https://rolflekang.com/serving-plain-text-with-nextjs",
      "link": "https://rolflekang.com/serving-plain-text-with-nextjs",
      "pubDate": "2020-05-26T00:00:00.000Z",
      "title": "Serving text/plain for curl with Next",
    },
    Object {
      "author": null,
      "categories": Array [
        "networking",
        "unifi",
        "ubiquiti",
        "homelab",
      ],
      "description": null,
      "guid": "https://rolflekang.com/wireless-uplinks-with-unifi",
      "link": "https://rolflekang.com/wireless-uplinks-with-unifi",
      "pubDate": "2020-05-03T00:00:00.000Z",
      "title": "Wireless uplinks with Unifi",
    },
    Object {
      "author": null,
      "categories": Array [
        "git",
      ],
      "description": null,
      "guid": "https://rolflekang.com/using-git-commits-instead-of-stash",
      "link": "https://rolflekang.com/using-git-commits-instead-of-stash",
      "pubDate": "2019-11-14T00:00:00.000Z",
      "title": "Using git commits instead of git stash",
    },
    Object {
      "author": null,
      "categories": Array [
        "gatsby",
        "canvas",
        "javascript",
      ],
      "description": null,
      "guid": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
      "link": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
      "pubDate": "2019-02-26T00:00:00.000Z",
      "title": "Twitter cards for Gatsby posts",
    },
    Object {
      "author": null,
      "categories": Array [
        "gatsby",
        "javascript",
      ],
      "description": null,
      "guid": "https://rolflekang.com/related-articles-with-gatsby",
      "link": "https://rolflekang.com/related-articles-with-gatsby",
      "pubDate": "2019-02-20T00:00:00.000Z",
      "title": "Related articles with Gatsby",
    },
    Object {
      "author": null,
      "categories": Array [
        "reasonml",
        "native",
      ],
      "description": null,
      "guid": "https://rolflekang.com/creating-a-cli-with-reason-native",
      "link": "https://rolflekang.com/creating-a-cli-with-reason-native",
      "pubDate": "2019-02-12T00:00:00.000Z",
      "title": "Creating a CLI with Reason native",
    },
    Object {
      "author": null,
      "categories": Array [
        "graphql",
        "testing",
        "javascript",
      ],
      "description": null,
      "guid": "https://rolflekang.com/testing-simple-graphql-services",
      "link": "https://rolflekang.com/testing-simple-graphql-services",
      "pubDate": "2017-05-28T00:00:00.000Z",
      "title": "Testing simple GraphQL services",
    },
    Object {
      "author": null,
      "categories": Array [
        "running",
      ],
      "description": null,
      "guid": "https://rolflekang.com/lets-run-some-races",
      "link": "https://rolflekang.com/lets-run-some-races",
      "pubDate": "2017-01-02T00:00:00.000Z",
      "title": "Let's run some races this year",
    },
    Object {
      "author": null,
      "categories": Array [
        "ipad",
        "development-environment",
      ],
      "description": null,
      "guid": "https://rolflekang.com/software-development-on-an-ipad",
      "link": "https://rolflekang.com/software-development-on-an-ipad",
      "pubDate": "2017-01-01T00:00:00.000Z",
      "title": "Software development on an iPad",
    },
    Object {
      "author": null,
      "categories": Array [
        "linter",
        "linting",
        "development",
      ],
      "description": null,
      "guid": "https://rolflekang.com/filtering-lint-errors",
      "link": "https://rolflekang.com/filtering-lint-errors",
      "pubDate": "2016-06-10T00:00:00.000Z",
      "title": "Filtering lint errors",
    },
    Object {
      "author": null,
      "categories": Array [
        "travel",
        "interrail",
      ],
      "description": null,
      "guid": "https://rolflekang.com/interrail-summer",
      "link": "https://rolflekang.com/interrail-summer",
      "pubDate": "2015-07-26T00:00:00.000Z",
      "title": "Interrail summer",
    },
    Object {
      "author": null,
      "categories": Array [
        "atom",
        "writing",
        "procrastination",
      ],
      "description": null,
      "guid": "https://rolflekang.com/writing-latex-in-atom",
      "link": "https://rolflekang.com/writing-latex-in-atom",
      "pubDate": "2015-05-23T00:00:00.000Z",
      "title": "Writing (latex) in Atom",
    },
    Object {
      "author": null,
      "categories": Array [
        "python",
        "testing",
        "mocking",
      ],
      "description": null,
      "guid": "https://rolflekang.com/rmoq",
      "link": "https://rolflekang.com/rmoq",
      "pubDate": "2014-12-30T00:00:00.000Z",
      "title": "rmoq - a request mock cache",
    },
    Object {
      "author": null,
      "categories": Array [
        "django",
        "nopassword",
        "django-nopassword",
        "authentication",
      ],
      "description": null,
      "guid": "https://rolflekang.com/django-nopassword-one-point-o",
      "link": "https://rolflekang.com/django-nopassword-one-point-o",
      "pubDate": "2014-10-03T00:00:00.000Z",
      "title": "django-nopassword reached 1.0",
    },
    Object {
      "author": null,
      "categories": Array [
        "this",
        "static-site",
        "pelican",
        "jekyll",
      ],
      "description": null,
      "guid": "https://rolflekang.com/moving-to-pelican",
      "link": "https://rolflekang.com/moving-to-pelican",
      "pubDate": "2014-05-07T00:00:00.000Z",
      "title": "Moving the blog to pelican",
    },
    Object {
      "author": null,
      "categories": Array [
        "gruntjs",
        "cookiecutter",
        "tumblr",
      ],
      "description": null,
      "guid": "https://rolflekang.com/building-tumblr-theme-with-grunt",
      "link": "https://rolflekang.com/building-tumblr-theme-with-grunt",
      "pubDate": "2014-02-02T00:00:00.000Z",
      "title": "Building a Tumblr theme with GruntJS",
    },
    Object {
      "author": null,
      "categories": Array [
        "db",
        "backup",
        "redis",
      ],
      "description": null,
      "guid": "https://rolflekang.com/backup-routines-for-redisdb",
      "link": "https://rolflekang.com/backup-routines-for-redisdb",
      "pubDate": "2013-10-24T00:00:00.000Z",
      "title": "Backup routine for Redis",
    },
    Object {
      "author": null,
      "categories": Array [
        "django",
        "custom user",
      ],
      "description": null,
      "guid": "https://rolflekang.com/django-app-custom-user",
      "link": "https://rolflekang.com/django-app-custom-user",
      "pubDate": "2013-05-22T00:00:00.000Z",
      "title": "Making your Django app ready for custom users",
    },
    Object {
      "author": null,
      "categories": Array [
        "pypstats",
        "pypi",
        "stats",
        "geek-tool",
      ],
      "description": null,
      "guid": "https://rolflekang.com/stats-are-fun",
      "link": "https://rolflekang.com/stats-are-fun",
      "pubDate": "2013-03-03T00:00:00.000Z",
      "title": "Stats are fun",
    },
    Object {
      "author": null,
      "categories": Array [
        "alfredapp",
        "jekyll",
        "shell script",
      ],
      "description": null,
      "guid": "https://rolflekang.com/alfred-scripts-for-jekyll",
      "link": "https://rolflekang.com/alfred-scripts-for-jekyll",
      "pubDate": "2013-02-26T00:00:00.000Z",
      "title": "Alfred scripts for Jekyll",
    },
    Object {
      "author": null,
      "categories": Array [
        "tags",
        "postgres",
        "array",
        "pgarray",
        "django",
      ],
      "description": null,
      "guid": "https://rolflekang.com/use-array-in-postgresql",
      "link": "https://rolflekang.com/use-array-in-postgresql",
      "pubDate": "2013-01-27T00:00:00.000Z",
      "title": "Use array in postgresql with django",
    },
    Object {
      "author": null,
      "categories": Array [
        "puppet",
        "web.py",
        "nginx",
        "uwsgi",
        "github",
        "github-hooks",
      ],
      "description": null,
      "guid": "https://rolflekang.com/apply-puppet-configuration-automatically",
      "link": "https://rolflekang.com/apply-puppet-configuration-automatically",
      "pubDate": "2013-01-03T00:00:00.000Z",
      "title": "Apply puppet automatically",
    },
    Object {
      "author": null,
      "categories": Array [
        "wikipedia",
        "css",
        "readability",
        "stylebot",
      ],
      "description": null,
      "guid": "https://rolflekang.com/readable-wikipedia",
      "link": "https://rolflekang.com/readable-wikipedia",
      "pubDate": "2012-12-02T00:00:00.000Z",
      "title": "Readable Wikipedia",
    },
    Object {
      "author": null,
      "categories": Array [
        "django",
        "no-password",
        "email-authentication",
        "authentication",
      ],
      "description": null,
      "guid": "https://rolflekang.com/django-nopassword",
      "link": "https://rolflekang.com/django-nopassword",
      "pubDate": "2012-09-28T00:00:00.000Z",
      "title": "django-nopassword",
    },
    Object {
      "author": null,
      "categories": Array [
        "arduino",
        "github",
        "git",
        "web.py",
      ],
      "description": null,
      "guid": "https://rolflekang.com/the-github-lamp",
      "link": "https://rolflekang.com/the-github-lamp",
      "pubDate": "2012-09-16T00:00:00.000Z",
      "title": "The Github lamp",
    },
  ],
  "feedLink": undefined,
  "link": "https://rolflekang.com",
  "parser": "FEEDPARSER",
  "title": "Writing by Rolf Erik Lekang",
}
`;
